pennylane/labs/intermediate_reps/rowcol.py

```diff
@@ -0,0 +1,214 @@
+        source_parent (Optional[int]): Parent node of ``source`` in ``tree``.
+
+    Returns:
+        list[tuple[int]]: Pairs of nodes that constitute post-order traversal of ``tree``
+        starting at ``source``. Strictly speaking, the traversal is encoded in the first
+        entry of each pair, and the second entry simply is the parent node for each first entry.
+
+    A useful illustration of depth-first tree traversals can be found on
+    `Wikipedia <https://en.wikipedia.org/wiki/Tree_traversal#Depth-first%20search>`__.
+    """
+    out = []
+    # First, traverse the subtrees attached to the current node ``source``
+    for child in tree.neighbors(source):
+        # The graph does not know of the tree structure, so we need to make sure we skip
+        # the parent of ``source`` among its neighbors.
+        if child != source_parent:
+            # Recurse
+            out.extend(postorder_traverse(tree, child, source))
+    # Second, attach the current node itself, together with its parent
+    if source_parent is not None:
+        out.append((source, source_parent))
+
+    return out
+
+
+def preorder_traverse(tree: nx.Graph, source: int, source_parent: int = None):
+    """Pre-order traverse a tree graph, starting from (but excluding) the node ``source``.
+
+    Args:
+        tree (nx.Graph): Tree graph to traverse. Must contain ``source``. Must contain
+            ``source_parent`` if it is not None. Typing assumes integer-labeled nodes.
+        source (int): Node to start the traversal from
+        source_parent (Optional[int]): Parent node of ``source`` in ``tree``.
+
+    Returns:
+        list[tuple[int]]: Pairs of nodes that constitute pre-order traversal of ``tree``
+        starting at ``source``. Strictly speaking, the traversal is encoded in the first
+        entry of each pair, and the second entry simply is the parent node for each first entry.
+
+    A useful illustration of depth-first tree traversals can be found on
+    `Wikipedia <https://en.wikipedia.org/wiki/Tree_traversal#Depth-first%20search>`__.
+    """
+    out = []
+
+    # First, attach the current node itself, together with its parent
+    if source_parent is not None:
+        out.append((source, source_parent))
+    # Second, traverse the subtrees attached to the current node ``source``
+    for child in tree.neighbors(source):
+        # The graph does not know of the tree structure, so we need to make sure we skip
+        # the parent of ``source`` among its neighbors.
+        if child != source_parent:
+            # Recurse
+            out.extend(preorder_traverse(tree, child, source))
+
+    return out
+
+
+def _update(P: np.ndarray, cnots: list[tuple[int]], control: int, target: int):
+    """In-place apply update corresponding to a CNOT on wires ``(control, target)``
+    to parity matrix P and list of CNOT gates ``cnots``."""
+    P[target] += P[control]
+    cnots.append((control, target))
```

Comment:
a bit sketchy with no returns 😬 can this be fixed or would it make things significantly more complicated?